palace/utils/geodata.cpp

```diff
@@ -1073,6 +1140,62 @@ void GetSurfaceNormal(mfem::ParMesh &mesh, const mfem::Array<int> &marker,
   // }
 }
 
+void RebalanceConformalMesh(std::unique_ptr<mfem::ParMesh> &mesh, const IoData &iodata,
+                            const std::string &output_serial_mesh_file)
+{
+  // Write the parallel mesh to a stream as a serial mesh, then read back in and partition
+  // using METIS.
+  auto comm = mesh->GetComm();
+  constexpr bool generate_bdr = false, generate_edges = true, refine = true,
+                 fix_orientation = true;
+  const bool save_serial_mesh = !output_serial_mesh_file.empty();
+  std::unique_ptr<mfem::Mesh> smesh;
+  std::unique_ptr<int[]> partitioning;
+  if constexpr (false)
+  {
+    // Write the serial mesh to a stream and read that through the Mesh constructor.
+    std::stringstream msg;
+    msg.precision(mesh::MSH_FLT_PRECISION);
+    mesh->PrintAsSerial(msg);
+    mesh.reset();  // Release the no longer needed memory
+    if (Mpi::Root(comm))
+    {
+      smesh = std::make_unique<mfem::Mesh>(msg, generate_edges, refine, fix_orientation);
+    }
+  }
+  else
+  {
+    // Directly ingest the generated Mesh.
+    smesh = std::make_unique<mfem::Mesh>(mesh->GetSerialMesh(0));
+    mesh.reset();  // Release the no longer needed memory
+    if (Mpi::Rank(comm) != 0)
+    {
+      smesh.reset();
+    }
+  }
+
+  if (Mpi::Root(comm))
+  {
+    smesh->FinalizeTopology(generate_bdr);
+    smesh->Finalize(refine, fix_orientation);
+    partitioning = GetMeshPartitioning(*smesh, Mpi::Size(comm));
+  }
+
+  mesh = DistributeMesh(comm, smesh, partitioning);
+  if (save_serial_mesh)
+  {
+    BlockTimer bt(Timer::IO);
+    if (Mpi::Root(comm))
+    {
+      std::ofstream serial(output_serial_mesh_file);
+      serial.precision(mesh::MSH_FLT_PRECISION);
+      mesh::DimensionalizeMesh(*smesh, iodata.GetMeshScaleFactor());
+      smesh->Print(serial);  // Do not need to nondimensionalize the temporary
+    }
+    Mpi::Barrier(comm);
```

Comment:
Do we need a barrier here?

Reply:
The barrier here is to match with the `BlockTimer bt(Timer::IO)` before the branch on` Mpi::Root(comm)`. I copied the pattern in `ReadMesh` where the timer is instantiated for all ranks rather than just root, and kept coordinated by the barrier. Strictly speaking the other ranks don't need to be waiting here, then the timer could be moved inside the `if`, and only time the root processor, and the IO timing would then be desynchronized across processors.

Reply:
Whoops sorry about that that, my mistake. Not sure how I missed the `Timer::IO` addition already as well.